File: Node/simple-root-bot/rootBot.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ActivityHandler, ActionTypes, ActivityTypes, CardFactory, MessageFactory  } = require('botbuilder');

class RootBot extends ActivityHandler {
    constructor(conversationState, skillsConfig, skillClient) {
        super();
        if (!conversationState) throw new Error('[RootBot]: Missing parameter. conversationState is required');
        if (!skillsConfig) throw new Error('[RootBot]: Missing parameter. skillsConfig is required');
        if (!skillClient) throw new Error('[RootBot]: Missing parameter. skillClient is required');

        this.conversationState = conversationState;
        this.skillsConfig = skillsConfig;
        this.skillClient = skillClient;

        this.botId = process.env.MicrosoftAppId;
        if (!this.botId) {
            throw new Error('[RootBot] MicrosoftAppId is not set in configuration');
        }

        // Create state property to track the active skill
        this.activeSkillProperty = this.conversationState.createProperty('activeSkillProperty');

        // See https://aka.ms/about-bot-activity-message to learn more about the message and other activity types.
        this.onMessage(async (context, next) => {
            // Try to get the active skill
            let activeSkill = await this.activeSkillProperty.get(context);

            if (activeSkill) {
                // Send the activity to the skill
                await this.sendToSkill(context, activeSkill);
            } else {

                const text = context.activity.text.toLowerCase();
                if (text.includes('dotnet') || text.includes('javascript') || text.includes('python')) {
                    activeSkill = this.skillsConfig.skills[text];
                    await context.sendActivity('Got it, connecting you to the skill...');

                    // Set active skill
                    await this.activeSkillProperty.set(context, activeSkill);

                    // Send the activity to the skill
                    await this.sendToSkill(context, activeSkill);
                } else {
                    const card = this.getOptionsCard();
                    const message = MessageFactory.attachment(card);
                    await context.sendActivity(message);
                }
            }

            // By calling next() you ensure that the next BotHandler is run.
            await next();
        });

        this.onUnrecognizedActivityType(async (context, next) => {
            // Handle EndOfConversation returned by the skill.
            if (context.activity.type === ActivityTypes.EndOfConversation) {
                // Stop forwarding activities to Skill.
                await this.activeSkillProperty.set(context, undefined);

                // Show status message, text and value returned by the skill
                let eocActivityMessage = `Received ${ ActivityTypes.EndOfConversation }.\n\nCode: ${ context.activity.code }`;
                if (context.activity.text) {
                    eocActivityMessage += `\n\nText: ${ context.activity.text }`;
                }

                if (context.activity.value) {
                    eocActivityMessage += `\n\nValue: ${ context.activity.value }`;
                }

                await context.sendActivity(eocActivityMessage);

                // We are back at the root
                await context.sendActivity('Back in the root javascript bot.');
                const card = this.getOptionsCard();
                const message = MessageFactory.attachment(card);
                await context.sendActivity(message);

                // Save conversation state
                await this.conversationState.saveChanges(context, true);
            }

            // By calling next() you ensure that the next BotHandler is run.
            await next();
        });

        this.onMembersAdded(async (context, next) => {
            const membersAdded = context.activity.membersAdded;
            for (let cnt = 0; cnt < membersAdded.length; ++cnt) {
                if (membersAdded[cnt].id !== context.activity.recipient.id) {
                    await context.sendActivity('Hello and welcome!');
                }
            }

            // By calling next() you ensure that the next BotHandler is run.
            await next();
        });

        this.onDialog(async (context, next) => {
            // Save any state changes. The load happened during the execution of the Dialog.
            await this.conversationState.saveChanges(context);

            // By calling next() you ensure that the next BotHandler is run.
            await next();
        });
    }

    getOptionsCard() {
        const buttons = [
            { type: ActionTypes.ImBack, title: 'Dotnet Skill', value: 'dotnet' },
            { type: ActionTypes.ImBack, title: 'Javascript Skill', value: 'javascript' },
            { type: ActionTypes.ImBack, title: 'Python Skill', value: 'python' },
            { type: ActionTypes.ImBack, title: 'Dotnet V3 Skill', value: 'dotnetv3' }
        ];

        const card = CardFactory.heroCard('Javascript Skills Bot Options', undefined,
            buttons, { text: 'You can upload an image or select one of the following choices.' });

        return card;
    }

    async sendToSkill(context, targetSkill) {
        // NOTE: Always SaveChanges() before calling a skill so that any activity generated by the skill
        // will have access to current accurate state.
        await this.conversationState.saveChanges(context, true);

        // route the activity to the skill
        const response = await this.skillClient.postToSkill(this.botId, targetSkill, this.skillsConfig.skillHostEndpoint, context.activity);

        // Check response status
        if (!(response.status >= 200 && response.status <= 299)) {
            throw new Error(`[RootBot]: Error invoking the skill id: "${ targetSkill.id }" at "${ targetSkill.skillEndpoint }" (status is ${ response.status }). \r\n ${ response.body }`);
        }
    }
}

module.exports.RootBot = RootBot;

File: Python/simple-root-bot/bots/root_bot.py
from typing import List

from botbuilder.core import (
    ActivityHandler,
    BotFrameworkHttpClient,
    CardFactory,
    ConversationState,
    MessageFactory,
    TurnContext,
)

from botbuilder.schema import (
    Attachment,
    ActivityTypes,
    ChannelAccount,
    CardAction,
    HeroCard,
)

from config import DefaultConfig, SkillConfiguration

ACTIVE_SKILL_PROPERTY_NAME = "activeSkillProperty"

class RootBot(ActivityHandler):
    def __init__(
        self,
        conversation_state: ConversationState,
        skills_config: SkillConfiguration,
        skill_client: BotFrameworkHttpClient,
        config: DefaultConfig,
    ):
        self._bot_id = config.APP_ID
        self._skill_client = skill_client
        self._skills_config = skills_config
        self._conversation_state = conversation_state
        self._active_skill_property = conversation_state.create_property(
            "activeSkillProperty"
        )

    async def on_turn(self, turn_context: TurnContext):
        if turn_context.activity.type == ActivityTypes.end_of_conversation:
            # Handle end of conversation back from the skill
            # forget skill invocation
            await self._active_skill_property.delete(turn_context)
            await self._conversation_state.save_changes(turn_context, force=True)

            # We are back
            await turn_context.send_activity(
                MessageFactory.text("Hi. Back in the root bot.")
            )

            await turn_context.send_activity(
                MessageFactory.attachment(CardFactory.hero_card(self._get_hero_card()))
            )
        else:
            await super().on_turn(turn_context)

    async def on_message_activity(self, turn_context: TurnContext):
        # If there is an active skill
        active_skill: str = await self._active_skill_property.get(turn_context)

        if active_skill:
            # NOTE: Always SaveChanges() before calling a skill so that any activity generated by the skill
            # will have access to current accurate state.
            await self._conversation_state.save_changes(turn_context, force=True)

            # route activity to the skill
            await self._skill_client.post_activity(
                self._bot_id,
                active_skill,
                self._skills_config.SKILL_HOST_ENDPOINT,
                turn_context.activity,
            )
        else:
            text = turn_context.activity.text
            if "dotnet" in text or "javascript" in text or "python" in text:
                await turn_context.send_activity(
                    MessageFactory.text("Got it, connecting you to the skill...")
                )

                active_skill = self._skills_config.SKILLS[text]
                # save ConversationReferene for skill
                await self._active_skill_property.set(turn_context, active_skill)

                # NOTE: Always SaveChanges() before calling a skill so that any activity generated by the
                # skill will have access to current accurate state.
                await self._conversation_state.save_changes(turn_context, force=True)

                await self._skill_client.post_activity(
                    self._bot_id,
                    active_skill,
                    self._skills_config.SKILL_HOST_ENDPOINT,
                    turn_context.activity,
                )
            else:
                # just respond
                await turn_context.send_activity(MessageFactory.text("Me no nothin'."))
                await turn_context.send_activity(
                    MessageFactory.attachment(
                        CardFactory.hero_card(self._get_hero_card())
                    )
                )

    async def on_members_added_activity(
        self, members_added: List[ChannelAccount], turn_context: TurnContext
    ):
        for member in members_added:
            if member.id != turn_context.activity.recipient.id:
                await turn_context.send_activity(
                    MessageFactory.text("Hello and welcome!")
                )

    def _get_hero_card(self) -> Attachment:
        return HeroCard(
            title="Python Skills Bot Options",
            text="Click one of the buttons below to initiate that echo skill.",
            buttons=[
                CardAction(
                    type="imBack", title="Dotnet Skill", text="dotnet", value="dotnet"
                ),
                CardAction(
                    type="imBack",
                    title="Javascript Skill",
                    text="javascript",
                    value="javascript",
                ),
                CardAction(
                    type="imBack", title="Python Skill", text="python", value="python"
                ),
                CardAction(
                    type="imBack", title="Dotnet V3 Skill", text="dotnetv3", value="dotnetv3"
                ),
            ],
        )